### PR TITLE
fix: Fix cors headers

### PR DESCRIPTION
**Pull request by alexgarel**
* centralize cors headers handling in perl code
* add Access-Control-Allow-Methods correctly
* add tests

* cleaned nginx prod configurations
* make docker nginx more close to production configuration

should fix:
- #7796
- #7901

### CODE REVIEW

**Comment by teolemon** [comment]
Noice 👍

**Review by teolemon — APPROVED** [review]

**Comment by alexgarel** [comment]
I want a validation of @stephanegigandet on this one, as it is a big refactor !

File: conf/nginx.conf
Comment (stephanegigandet) [inline conf/nginx.conf:57]: It looks like it's included twice.

**Comment by stephanegigandet** [comment]
Looks good to me, but I have not tested it. Maybe wait until January until we merge and deploy this.

**Comment by stephanegigandet** [comment]
Thanks for adding all the tests. :)

File: conf/nginx.conf
Comment (stephanegigandet) [inline conf/nginx.conf:14]: I think we should move this file to conf/nginx/nginx.conf (which currently seems to contain the default nginx.conf)

File: conf/nginx/sites-available/off
Comment (stephanegigandet) [inline conf/nginx/sites-available/off:1]: What does this line do? And is there a reason why it is in the off file but not obf etc.?

File: conf/nginx.conf
Comment (stephanegigandet) [reply conf/nginx.conf:14]: ah I'm confused, this nginx.conf file is only for docker, which does not use the files in sites-available?

File: conf/nginx/sites-available/off
Comment (stephanegigandet) [reply conf/nginx/sites-available/off:1]: isn't it something that belongs in another file, e.g. gulpfile.js?

**Comment by stephanegigandet** [comment]
Deployed on https://world.openfoodfacts.dev and https://world.pro.openfoodfacts.dev with a few minor changes to add the off:off authentication and change .org to .dev

also removed the const { off } = require("gulp") line

**Comment by stephanegigandet** [comment]
For some reason, I don't get the access-control-allow-origin: * header for v3 API queries, but I do get it for v2. (same in prod in fact)

```
$ curl -i -X OPTIONS https://off:off@world.openfoodfacts.dev/api/v3/product/12345678143
HTTP/2 404 
server: nginx/1.10.3
date: Mon, 02 Jan 2023 17:55:09 GMT
content-type: application/json; charset=utf-8
x-frame-options: DENY
x-content-type-options: nosniff
x-download-options: noopen
x-xss-protection: 1; mode=block
x-request-id: cJ90QvVhm9SHiqqk

{"code":"12345678143","errors":[{"field":{"id":"12345678143","value":"12345678143"},"impact":{"id":"failure","lc_name":"Failure","name":"Failure"},"message":{"id":"product_not_found","lc_name":"","name":""}}],"result":{"id":"product_not_found","lc_name":"Product not found","name":"Product not found"},"status":"failure","warnings":[]}

$ curl -i -X OPTIONS https://off:off@world.openfoodfacts.dev/api/v2/product/12345678143
HTTP/2 404 
server: nginx/1.10.3
date: Mon, 02 Jan 2023 17:55:17 GMT
content-type: application/json; charset=utf-8
access-control-allow-origin: *
x-frame-options: DENY
x-content-type-options: nosniff
x-download-options: noopen
x-xss-protection: 1; mode=block
x-request-id: kP8kuV7uX2Gq5ZkN
```

File: conf/nginx.conf
Comment (alexgarel) [reply conf/nginx.conf:14]: yes for docker we have a different config than in prod.

So I kept the nginx conf for prod in conf/nginx, while docker simply uses conf/nginx.conf
(a docker nginx instance only serve one application, and obf, opf, etc. are not docker ready yet.)

File: conf/nginx/sites-available/off
Comment (alexgarel) [reply conf/nginx/sites-available/off:1]: I don't know myself why it's there :-D

Comment (alexgarel) [reply conf/nginx/sites-available/off:1]: Will remove in next commit.

**Comment by alexgarel** [comment]
@stephanegigandet 
> For some reason, I don't get the access-control-allow-origin: * header for v3 API queries, but I do get it for v2. (same in prod in fact)

Are you sure you have the right code ? (it's not just a nginx config)

For you should not only have "access-control-allow-origin: *" but also the "Access-Control-Allow-Methods" header.

**Comment by stephanegigandet** [comment]
I redeployed and tested again, all good!

**Review by stephanegigandet — APPROVED** [review]

**Pull request merged**
